### PR TITLE
Spec for spender/expenses POST

### PR DESCRIPTION
**Pull request by kartikeyrajvaidya**

### CODE REVIEW

File: src/components/schemas/expense.yaml [hunk 1]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+  type: object
+  additionalProperties: false
+  properties:
+    id:
+      $ref: './fields.yaml#/id_string'
+    spent_at:
+      allOf:
+        - $ref: './fields.yaml#/timestamptz_pst'
+      description: |
+        signifies when this expense was made.
```
[the remark this answers is not in the document]
Comment (viiicky) [reply src/components/schemas/expense.yaml:543]: or start as `It signifies...`

File: src/components/schemas/expense.yaml [hunk 2]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+        This date cannot be in future.
+    source:
+      type: string
+      description: |
+        The name of the client app which is creating the expense.
+        if its a thrid party app, it prefixed with TPA_.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:549]: i capital in if

File: src/components/schemas/expense.yaml [hunk 3]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+        Ex - If you pass SLACK_APP, it will be saved as TPA_SLACK_APP. If not
+        passed, source is saved as TPA.
+      example: SLACK_APP
+    merchant:
+      type: string
+      maxLength: 255
+      nullable: true
+      description: |
+        String denoting the name of the vendor to which the Expense was made.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:558]: can consider expense instead of Expense

File: src/components/schemas/expense.yaml [hunk 4]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+      description: Tax amount as setup by your organization.
+    tax_group_id:
+      allOf:
+        - $ref: './fields.yaml#/fk_string'
+      nullable: true
+    is_billable:
+      type: boolean
+      nullable: true
+      example: true
+      description: This fields marks whether a given expense is billable a not.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:598]: fields -> field

File: src/components/schemas/expense.yaml [hunk 5]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+    is_reimbursable:
+      type: boolean
+      nullable: true
+      description: This fields marks whether a given expense is reimbursable a not.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:602]: fields -> field

File: src/components/schemas/expense.yaml [hunk 6]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+        - MILES
+        - null
+      example: MILES
+    started_at:
+      $ref: './fields.yaml#/timestamptz_pst'
+    ended_at:
+      $ref: './fields.yaml#/timestamptz_pst'
+    travel_classes:
+      description: |
+        This field is applicable only for `Flight`, `Train`, `Bus`, `Taxi` category.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:623]: can you confirm if it's true for `Taxi` category also. Asking cause I recently checked the definition of sync function while working on policy doc, and I remember only three categories being handled - flight, train and bus

File: src/components/schemas/expense.yaml [hunk 7]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+        index 0 holds onward journey class, index 1 holds return journey travel class.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:624]: Start with capital I

Comment (viiicky) [reply src/components/schemas/expense.yaml:624]: either make `journey class` at both the places or make `journey travel class` at both the places

File: src/components/schemas/expense.yaml [hunk 8]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+      items:
+        type: string
+      example:
+        - ECONOMY
+        - BUSINESS
+    hotel_is_breakfast_provided:
+      type: boolean
+      nullable: true
+      description: |
+        This field is applicable for only `Hotel` category.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:637]: for only -> only for

File: src/components/schemas/expense.yaml [hunk 9]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+    locations:
+      description: |
+        This field will be present for travel related categories.
+        For travel between Place A to Place B, this array will contain two items, From location at array index 0 and To location at array index 1.
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:641]: items, -> items.

File: src/components/schemas/expense.yaml [hunk 10]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+        For travel between Place A to Place c and stopping at Place B, this
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:642]: incomplete sentence

File: src/components/schemas/expense.yaml [hunk 11]
```diff
@@ -531,13 +531,122 @@ expense_in:
...
+      type: array
+      minItems: 0
+      maxItems: 10
```
Comment (viiicky) [inline src/components/schemas/expense.yaml:645]: dont know about this; do we restrict to 10 locations?

File: reference/spender.yaml [hunk 12]
```diff
@@ -1320,6 +1320,143 @@ components:
           nullable: true
           description: |
             signifies when this expense was last settled.
+    timestamptz_pst:
```
Comment (viiicky) [inline reference/spender.yaml:1323]: all the same comments in this green diff section as below

**Review by viiicky — CHANGES_REQUESTED** [review]
added comments

File: src/components/schemas/fields.yaml [hunk 13]
```diff
@@ -128,6 +128,80 @@ category_ids:
     type: integer
     example: 16570
 
+locations:
+  description: |
+    This field will be present for travel related categories.
+    For travel between Place A to Place B, this array will contain two items, From location at array index 0 and To location at array index 1.
```
Comment (aditya999123) [inline src/components/schemas/fields.yaml:134]: index i and to localtion at index i+1

File: src/components/schemas/fields.yaml [hunk 14]
```diff
@@ -128,6 +128,80 @@ category_ids:
...
+    - BUSINESS
+
+distance_unit:
+  type: string
+  nullable: true
+  description: Unit of distance travelled.
+  enum:
+    - KM
+    - MILES
+    - null
```
Comment (aditya999123) [inline src/components/schemas/fields.yaml:161]: nullable as true, enum should not have null

File: src/components/schemas/expense.yaml [hunk 15]
```diff
@@ -531,13 +531,69 @@ expense_in:
...
 
 spender_expense_in:
-  allOf:
-    - type: object
-      required: [ 'claim_amount' ]
-      properties:
-        claim_amount:
-          $ref: './fields.yaml#/amount'
-    - $ref: '#/expense_in'
+  type: object
```
Comment (aditya999123) [inline src/components/schemas/expense.yaml:534]: add required keys

**Review by aditya999123 — APPROVED** [review]

File: reference/spender.yaml [hunk 16]
```diff
@@ -1329,6 +1331,149 @@ components:
...
+      description: Unit of distance travelled.
+      enum:
+        - KM
+        - MILES
+      example: MILES
+    travel_classes:
+      description: >
+        This field is applicable only for `Flight`, `Train`, `Bus` category.
+
+        index 0 holds onward journey travel class, index 1 holds return journey
```
Comment (viiicky) [inline reference/spender.yaml:1386]: Capital I

File: reference/spender.yaml [hunk 17]
```diff
@@ -1329,6 +1331,149 @@ components:
...
+          nullable: true
+          example: 92895
+        source_account_id:
+          allOf:
+            - $ref: '#/components/schemas/fk_string'
+          example: acwbl222wlg
+        tax_amount:
+          type: number
+          nullable: true
+          description: Tax amount as setup by your organization.
```
Comment (viiicky) [inline reference/spender.yaml:1454]: hmmm, tax_amount is setup by organisation??

File: src/components/schemas/fields.yaml [hunk 18]
```diff
@@ -135,6 +135,78 @@ category_ids:
...
+  nullable: true
+  description: |
+    String denoting the name of the vendor to which the expense was made.
+  example: Uber
+
+is_billable:
+  type: boolean
+  nullable: true
+  example: true
+  description: This field marks whether a given expense is billable a not.
```
Comment (viiicky) [inline src/components/schemas/fields.yaml:196]: a -> or
please do a round of self review before raising the PR

File: src/components/schemas/fields.yaml [hunk 19]
```diff
@@ -135,6 +135,78 @@ category_ids:
...
+  
+is_reimbursable:
+  type: boolean
+  nullable: true
+  description: This field marks whether a given expense is reimbursable a not.
```
Comment (viiicky) [inline src/components/schemas/fields.yaml:201]: a -> or
please do a round of self review before raising the PR

File: reference/spender.yaml (hunk 16 above)
Comment (viiicky) [reply reference/spender.yaml:1386]: fix it everywhere applicable

**Pull request merged**